[vulkan] Add, sub, mul, and div ops with broadcasting for Vulkan

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52842 [vulkan] Add, sub, mul, and div ops with broadcasting for Vulkan**

Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

Commit:
[vulkan] Add, sub, and mul with broadcasting ops for Vulkan

[ghstack-poisoned]

Commit:
Update on "[vulkan] Add, sub, and mul with broadcasting ops for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Add, sub, and mul with broadcasting ops for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Add, sub, and mul with broadcasting ops for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Add, sub, and mul with broadcasting ops for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Add, sub, mul, and div ops with broadcasting for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/add.glsl:
```diff
@@ -9,7 +9,9 @@ layout(set = 0, binding = 0) uniform PRECISION restrict writeonly image3D   uOut
+  ivec4 size;
+  ivec4 isize0;
+  ivec4 isize1;
   float alpha;
```
Can this be packed more tightly?  I'm thinking:

ivec3 size
float alpha
ivec4 isize0
ivec4 isize1

If this works, same for others in this PR as well.

SS-JIA replied:
I'll check again, but I think there were some issues with the alignment (as we've seen before when using vec3)

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/add_.glsl:
```diff
@@ -8,7 +8,8 @@ layout(std430) buffer;
-  ivec3 size;
+  ivec4 size;
+  ivec4 isize;
   float alpha;
```
Same

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/div_scalar.glsl:
```diff
@@ -0,0 +1,26 @@
+        pos,
+        texelFetch(uInput, pos, 0) / uBlock.other);
+  }
+}
```
Remove this entire codepath and re-use multiply by (1.0f / other).  Not only will that have the benefit of code reuse, it will also reduce binary size and multiplication is also typically faster than division.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/div_scalar_.glsl:
```diff
@@ -0,0 +1,25 @@
+        pos,
+        imageLoad(uOutput, pos) / uBlock.other);
+  }
+}
```
Same as above.  Replace div with mul by (1.0f / factor)

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/sub_scalar.glsl:
```diff
@@ -0,0 +1,26 @@
+        pos,
+        texelFetch(uInput, pos, 0) - uBlock.other);
+  }
+}
```
Same as div.  Please replace sub_scalar with add_scalar + (-factor).

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/sub_scalar_.glsl:
```diff
@@ -0,0 +1,25 @@
+        pos,
+        imageLoad(uOutput, pos) - uBlock.other);
+  }
+}
```
Same as above.  Please replace sub_scalar_ with add_scalar_ + (-factor).

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+      (input2.extents().data[2u] > 1 && input1.extents().data[2u] == 1) ||
+      input2.extents().data[0u] > input1.extents().data[0u]) {
+    return true;
+  }
```
You can just return the "if" condition.  if ( condition ) return true, else return false is equivalent to return condition.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+    TORCH_CHECK(
+        channels_size(input1) % 4 == 0,
+        "Vulkan binary elementwise ops require channel to be a multiple of 4 to broadcast along batch dimension!")
+  }
```
Please consider using more vertical spacing in your functions.  It makes code easier to read.  I generally group related stuff together but try not to put more than four or five lines next to one another.  Makes reading the code a lot easier.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+  const uint32_t input1_h = height_size(input1);
+  const uint32_t input1_w = width_size(input1);
+  const uint32_t input2_h = height_size(input2);
+  const uint32_t input2_w = width_size(input2);
```
I commend your use of const!

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+  return self;
+}
+
+Tensor binary_elementwise_tensor(
```
Very nice refactoring.  I like it!

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+      const float alpha = alpha_arg ? alpha_arg->to<float>() : 1.0;
+      const struct Block final {
+        uvec3 extents;
+        uint32_t fill_0;
```
If you decide to pack tighter you can replace this with float alpha.  Completely up to you.  It's not going to be a performance boost and only a very minor performance saving.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/BinaryElementwise.cpp:
```diff
@@ -0,0 +1,432 @@
+
+  return self;
+}
+
```
I like all this refactoring.  Great call.

AshkanAliabadi reviewed (approved):
Excellent, thanks Stephen!

Commit:
Update on "[vulkan] Add, sub, mul, and div ops with broadcasting for Vulkan"


Differential Revision: [D26665326](https://our.internmc.facebook.com/intern/diff/D26665326)

[ghstack-poisoned]
